cabal-install/src/Distribution/Client/ScriptUtils.hs

```diff
@@ -0,0 +1,342 @@
+import Distribution.Simple.Flag
+         ( fromFlagOrDefault )
+import Distribution.Simple.Setup
+         ( Flag(..) )
+import Distribution.CabalSpecVersion (CabalSpecVersion (..), cabalSpecLatest)
+import Distribution.Verbosity
+         ( normal )
+import Distribution.Simple.Utils
+         ( warn, die', createDirectoryIfMissingVerbose
+         , createTempDirectory, handleDoesNotExist )
+import Distribution.Client.ProjectConfig
+         ( ProjectConfig(..), ProjectConfigShared(..)
+         , withProjectOrGlobalConfig )
+import Distribution.Client.ProjectFlags
+         ( flagIgnoreProject )
+import Distribution.Client.TargetSelector
+         ( TargetSelectorProblem(..), TargetString(..) )
+import Distribution.Client.Types
+         ( PackageLocation(..), PackageSpecifier(..), UnresolvedSourcePackage )
+import Distribution.FieldGrammar
+         ( takeFields, parseFieldGrammar )
+import Distribution.PackageDescription.FieldGrammar
+         ( executableFieldGrammar )
+import Distribution.PackageDescription.PrettyPrint
+         ( writeGenericPackageDescription, showGenericPackageDescription )
+import Distribution.Parsec
+         ( Position(..) )
+import Distribution.Fields
+         ( ParseResult, parseString, parseFatalFailure, readFields )
+import qualified Distribution.SPDX.License as SPDX
+import Distribution.Solver.Types.SourcePackage as SP
+         ( SourcePackage(..) )
+import Distribution.Types.BuildInfo
+         ( BuildInfo(..) )
+import Distribution.Types.CondTree
+         ( CondTree(..) )
+import Distribution.Types.Executable
+         ( Executable(..) )
+import Distribution.Types.GenericPackageDescription as GPD
+         ( GenericPackageDescription(..), emptyGenericPackageDescription )
+import Distribution.Types.PackageDescription
+         ( PackageDescription(..), emptyPackageDescription )
+import Distribution.Types.PackageName.Magic
+         ( fakePackageId )
+import Language.Haskell.Extension
+         ( Language(..) )
+import Distribution.Client.HashValue
+         ( hashValue, showHashValue )
+import Distribution.Simple.Utils
+         ( readUTF8File )
+
+import Control.Concurrent.MVar
+         ( newEmptyMVar, putMVar, tryTakeMVar )
+import Control.Exception
+         ( bracket )
+import qualified Data.ByteString.Char8 as BS
+import Data.ByteString.Lazy ()
+import qualified Text.Parsec as P
+import System.Directory
+         ( getTemporaryDirectory, removeDirectoryRecursive, doesFileExist, canonicalizePath )
+import System.FilePath
+         ( (</>), takeExtension )
+
+
+-- | Get the directory where script builds are cached.
+--
+-- CABAL_DIR\/script-builds\/
+getScriptCacheDirectoryRoot :: IO FilePath
+getScriptCacheDirectoryRoot = do
+  cabalDir <- getCabalDir
+  return $ cabalDir </> "script-builds"
+
+-- | Get the hash of a script path.
+getScriptHash :: String -> FilePath -> IO String
+getScriptHash prefix script = showHashValue . hashValue . fromString . (prefix ++) <$> canonicalizePath script
```

Comment:
Maybe relevant to note, at what point the hash changes (when the filename changes).

Reply:
I'll also make a note about prefixs, which I meant to do but apparently forgot.